x-pack/plugins/observability_solution/apm/server/deprecations/apm_user_role.ts

```diff
@@ -0,0 +1,179 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  SecurityGetRoleMappingResponse,
+  SecurityGetUserResponse,
+} from '@elastic/elasticsearch/lib/api/types';
+import type {
+  CoreSetup,
+  DeprecationsDetails,
+  DocLinksServiceSetup,
+  ElasticsearchClient,
+  GetDeprecationsContext,
+} from '@kbn/core/server';
+import { i18n } from '@kbn/i18n';
+import type { DeprecationApmDeps } from '.';
+import { deprecations } from '../lib/deprecations';
+
+const APM_USER_ROLE_NAME = 'apm_user';
+const getKibanaPrivilegesDocumentationUrl = (branch: string) => {
+  return `https://www.elastic.co/guide/en/kibana/${branch}/kibana-privileges.html`;
+};
+
+export async function getDeprecationsInfo(
+  { esClient }: GetDeprecationsContext,
+  core: CoreSetup,
+  apmDeps: DeprecationApmDeps
+) {
+  const client = esClient.asCurrentUser;
+  const { docLinks } = core;
+  const { security } = apmDeps;
+
+  // Nothing to do if security is disabled
+  if (!security?.license.isEnabled()) {
+    return [];
+  }
+
+  return [
+    ...(await getUsersDeprecations(client, apmDeps, docLinks)),
+    ...(await getRoleMappingsDeprecations(client, apmDeps, docLinks)),
+  ];
+}
+
+async function getUsersDeprecations(
+  client: ElasticsearchClient,
+  apmDeps: DeprecationApmDeps,
+  docLinks: DocLinksServiceSetup
+): Promise<DeprecationsDetails[]> {
+  const title = i18n.translate('xpack.apm.deprecations.apmUser.title', {
+    defaultMessage: `Check for users assigned the deprecated "{apmUserRoleName}" role`,
+    values: { apmUserRoleName: APM_USER_ROLE_NAME },
+  });
+
+  let users: SecurityGetUserResponse;
+  try {
+    users = await client.security.getUser();
+  } catch (err) {
+    const { logger } = apmDeps;
+    if (deprecations.getErrorStatusCode(err) === 403) {
+      logger.warn(
+        `Failed to retrieve users when checking for deprecations:` +
+          ` the "manage_security" cluster privilege is required.`
+      );
+    } else {
+      logger.error(
+        `Failed to retrieve users when checking for deprecations,` +
+          ` unexpected error: ${deprecations.getDetailedErrorMessage(err)}.`
+      );
+    }
+    return deprecations.deprecationError(title, err, docLinks);
+  }
+
+  const reportingUsers = Object.entries(users)
+    .filter(([_, user]) => user.roles.find(hasApmUserRole))
+    .map(([userName]) => userName);
+
+  if (reportingUsers.length === 0) {
+    return [];
+  }
+
+  return [
+    {
+      title,
+      message: i18n.translate('xpack.apm.deprecations.apmUser.description', {
+        defaultMessage: `The "{apmUserRoleName}" role has been deprecated. Remove the "{apmUserRoleName}" role from affected users in this cluster including: {users}`,
+        values: { apmUserRoleName: APM_USER_ROLE_NAME, users: reportingUsers.join(', ') },
```

Comment:
nit: by default `join()` separates the elements with a comma
```suggestion
        values: { apmUserRoleName: APM_USER_ROLE_NAME, users: reportingUsers.join() },
```